lms/djangoapps/lti_provider/outcomes.py

```diff
@@ -0,0 +1,199 @@
+"""
+Helper functions for managing interactions with the LTI outcomes service defined
+in LTI v1.1.
+"""
+
+from django.core.exceptions import ObjectDoesNotExist
+import logging
+from lxml import etree
+import requests
+import requests_oauthlib
+import uuid
+
+from lti_provider.models import GradedAssignment, LtiConsumer, OutcomeService
+
+log = logging.getLogger("edx.lti_provider")
+
+
+def store_outcome_parameters(request_params, user):
+    """
+    Determine whether a set of LTI launch parameters contains information about
+    an expected score, and if so create a GradedAssignment record. Create a new
+    OutcomeService record if none exists for the tool consumer, and update any
+    incomplete record with additional data if it is available.
+    """
+    result_id = request_params.get('lis_result_sourcedid', None)
+
+    # We're only interested in requests that include a lis_result_sourcedid
+    # parameter. An LTI consumer that does not send that parameter does not
+    # expect scoring updates for that particular request.
+    if result_id:
+        result_service = request_params.get('lis_outcome_service_url', None)
+        if not result_service:
+            # TODO: There may be a way to recover from this error; if we know
+            # the LTI consumer that the request comes from then we may be able
+            # to figure out the result service URL. As it stands, though, this
+            # is a badly-formed LTI request
+            log.warn(
+                "Outcome Service: lis_outcome_service_url parameter missing "
+                "from scored assignment; we will be unable to return a score. "
+                "Request parameters: %s",
+                request_params
+            )
+            return
+
+        # We need to identify which tool consumer made the request. The best way
+        # to do this is to use the tool_consumer_instance_guid, which should be
+        # unique for each tool consumer. Unfortunately, that LTI parameter is
+        # optional, so we can't rely on it being available. In that case, we can
+        # fall back on the consumer key as a means of identifying the consumer.
+        # Note that this depends on every consumer being assigned a unique key
+        # (which is best practice anyway).
+        instance_guid = request_params.get('tool_consumer_instance_guid', None)
+        consumer_key = request_params.get('oauth_consumer_key', None)
+        assert consumer_key, "oauth_consumer_key is not an optional parameter."
+
+        # Both usage and course ID parameters are supplied in the LTI launch URL
+        usage_key = request_params.get('usage_key', None)
+        course_key = request_params.get('course_key', None)
+        assert None not in [usage_key, course_key], "usage_key ({}) and course_key ({}) should not be None".format(
+            usage_key, course_key
+        )
```

Comment:
Please don't use asserts like exceptions. They're not guaranteed to run (e.g. if you use python -o). You can raise `ValueError` for this kind of thing.